### PR TITLE
DS-3137 Add Average feature

### PR DESCRIPTION
**Pull request by jrwishart**
- DS-3205 Add code to flatten tables and keep statistics [revdep skip]
- DS-3205 Improve utilities test coverage [ci skip]
- DS-3205 Update unit tests for new SumRows/Cols behaviour [revdep skip]
- DS-3205 Handle NULLs in check of multiple statistics [revdep skip]
- DS-3205 Fix bug in Matching of data.frames/Variable Sets in Sum [revdep skip]
- DS-3205 Version bump [ci skip]
- DS-3205 Correctly match columns when only single col exists [revdep skip]
- DS-3205 Fix column naming of edge case [revdep skip]
- DS-3205 Fix NAMESPACE for check note [revdep skip]
- DS-3205 Tidy up and robustify opposite Inf check [revdep skip]
- DS-3137 Remove call. argument [revdep skip]
- DS-3137 First draft of Average function
- DS-3137 Add more tests for Average [revdep skip]
- DS-3137 Amend name of attribute slightly
- DS-3137 Add draft of AverageColumns [revdep skip]
- DS-3137 Improve computation of sample size in Sum [revdep skip]
- fixup! DS-3137 Improve computation of sample size in Sum [revdep skip]
- DS-3137 Remove unused function [revdep skip]
- DS-3137 Fix tests after missing data retained during matching [revdep skip]
- DS-3137 Add helper function to compute number non missing obs [revdep skip]
- DS-3137 Refactor SumColumns [revdep skip]
- DS-3137 Ignore weights to Sum with multiple inputs [revdep skip]
- DS-3137 Refactor assignment of weights [revdep skip]
- DS-3137 Replace call stack code by passing an attribute [revdep skip]
- DS-3137 Add tests for AverageColumns [revdep skip]
- DS-3137 Remove old arg call. from documentation [revdep skip]
- DS-3137 Make warning of missing values more efficient [revdep skip]
- DS-3137 Refactor SumRows for AverageRows [revdep skip]
- DS-3137 First draft of Average Rows [revdep skip]
- DS-3137  Remove used functions and tidy up [revdep skip]
- DS-3137 Ensure args retained in AverageRows [revdep skip]
- DS-3137 Refactor counting for Averaging [revdep skip]
- DS-3137 Fix unit tests [revdep skip]
- DS-3137 Replace setRowNames with stats::setNames
- DS-3137 Don't name vectors when split from matrix [revdep skip]
- DS-3137 Don't give dimnames to reshaped scalars [revdep skip]
- DS-3137 Add AverageRows test file [revdep skip]
- DS-3137 Catch edge cases when splitting into 1D inputs [revdep skip]
- DS-3137 Improve documentation [revdep skip]
- DS-3137 Tweak example to avoid check note [revdep skip]
- DS-3137 Documentation update [revdep skip]
- DS-3137 Fix unit test [revdep skip]
- DS-3137 Don't name scalars but add names to Sum output [revdep skip]

### CODE REVIEW

**Review by mwmclean — CHANGES_REQUESTED** [review]
1. The documentation for `match.rows` in `Sum` mentions that unmatched rows should be NA or 0 when `match.rows` is `"YES"`, but it seems to just return the unmatched row/column in the output. E.g.
```
y <- matrix(1:12, nrow = 4, ncol = 3, dimnames = list(letters[1:4], letters[c(1, 2, 4)]))
x <- matrix(c(0, 1), byrow = TRUE, nrow = 4, ncol = 2, dimnames = list(letters[1:4], letters[c(1, 2)]))
verbs::Average(x, y)
```
2. Is there a reason you favour using this undocumented attribute to determine whether `Sum` has been called from `Average`? Clean Code types would strongly discourage it because it makes it so you need to read the code to understand what the function does. Why not an additional argument (that we don't include in the R code for the QScript) such as `append.sample.size.attribute`?

File: R/flatten.R [hunk 1]
```diff
@@ -59,8 +59,8 @@ FlattenTableAndDropStatisticsIfNecessary <- function(
                 table.out <- FlattenQTableToMatrix(table.out, 1:2, 3)
         }else if (n.dims == 4)  # e.g. Nominal - Multi by Binary - Gird
             table.out <- FlattenQTableToMatrix(table.out, c(1, 3), c(2, 4))
-##        else
-
+        else
+            table.out <- updateTableNamesWithColSpanLabels(table.out, table.out)
```
[the remark this answers is not in the document]
Comment (jrwishart) [reply R/flatten.R:63]: That was a mistake by me. I assume it occurred during the merge of the first/last change when there was merge conflicts. Should this diff block and the `## else` statement be deleted now?

File: R/sum.R [hunk 2]
```diff
@@ -3,7 +3,7 @@
 #'
 #' @description \code{Sum} is a generalization of \code{\link{sum}} and the + function in \code{\link{Arithmetic}}
 #'  but supporting additional pre-processing and matching of data before calculation.
-#' @param ... Objects to be summed; e.g. Q Tables, arrays, vectors, Variables or Variable Sets.
+#' @param ... Objects to be averaged; e.g. vectors, matrices, Variables, Variable Sets or Q Tables
```
[the remark this answers is not in the document]
Comment (jrwishart) [reply R/sum.R:6]: Mistake. I'll fix. Updated the wrong doc file when documenting `Average`

**Comment by jrwishart** [comment]
> 1. The documentation for `match.rows` in `Sum` mentions that unmatched rows should be NA or 0 when `match.rows` is `"YES"`, but it seems to just return the unmatched row/column in the output. E.g.
> 
> ```
> y <- matrix(1:12, nrow = 4, ncol = 3, dimnames = list(letters[1:4], letters[c(1, 2, 4)]))
> x <- matrix(c(0, 1), byrow = TRUE, nrow = 4, ncol = 2, dimnames = list(letters[1:4], letters[c(1, 2)]))
> verbs::Average(x, y)
> ```

The documentation mentions that when unmatched dimensions are found the resulting entries are either `NA` or `0` depending on the argument of `remove.missing`. The default of `remove.missing = TRUE` will result in the elements taking the value `0`. If `remove.missing = FALSE` then they will take the value of `NA`. So perhaps you were expecting the result of `verbs::Average(x, y, remove.missing = FALSE)` so that a column of `NA` values appears since column `d` doesn't exist in input element `x`?

> 
> 2. Is there a reason you favour using this undocumented attribute to determine whether `Sum` has been called from `Average`? Clean Code types would strongly discourage it because it makes it so you need to read the code to understand what the function does. Why not an additional argument (that we don't include in the R code for the QScript) such as `append.sample.size.attribute`?

The goal was to keep the arguments for `Sum` as parsimonious as possible. Adding another argument would go against that goal. The choices could be
* Add an argument to toggle the sample size information being returned in the output as an attribute as suggested. This has the benefit of not inspecting the attributes of the first input element to `Sum` and means `Average` would call `Sum` with that argument specified. However, it goes against the parsimonious goal that Tim wanted that there should be a minimal number of arguments in the function call.
* Keep it as an internal attribute specified in `Average` and will similarly appear in `Variance` and `StandardDeviation` (current approach). This means there isn't an argument to toggle its output behaviour. If a user wanted the same information, they could call `Sum` with the inputs inspected for missing data. E.g.

```
set.seed(3137)
in <- 1:12
in[sample(1:12)[1:3]] <- NA
y <- matrix(in, nrow = 4, ncol = 3, dimnames = list(letters[1:4], letters[c(1, 2, 4)]))
x <- matrix(sample(c(0, 1, NA), prob = c(2, 2, 1), replace = TRUE),
            byrow = TRUE, nrow = 4, ncol = 2, dimnames = list(letters[1:4], letters[c(1, 2)]))
x
y
Sum(x, y)
Sum(!is.na(x), !is.na(y))
```

* Always return the sample size information as an attribute to the output. This is probably the worst choice as it adds computation time that won't always be required.

I obviously favour the second option since it's included in this PR. Perhaps I could add further inline comments to make it easier for the code readers?

File: R/sum.R [hunk 3]
```diff
@@ -192,7 +200,18 @@ removeMissing <- function(x)
     x
 }
 
-isACallStartingWithSum <- function(x)
+#' @param sum.output The calculated output of the call to Sum just before it is returned.
+#' @param inputs integer of the number of inputs used in the call to Sum
```
[the remark this answers is not in the document]
Comment (jrwishart) [reply R/sum.R]: Fixed, updated the doc and removed obsolete variable in that function in 93445da3acc46df1054b9eaaba44e7b7a26ffe57

**Review by mwmclean — APPROVED** [review]
